torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1648,7 +1658,19 @@ struct to_ir {
         graph->createStore(x, fv)->insertBefore(false_block->return_node());
       }
 
-      auto unified = unifyTypes(tv->type(), fv->type());
+      // If the variable we're looking at is known to be Union[T1, T2],
+      // then it's okay to have one branch return T1 and the other
+      // return T2
+      SugaredValuePtr true_val = save_true->findInParentFrame(x);
+      SugaredValuePtr false_val = save_false->findInParentFrame(x);
+      bool default_to_union = false;
+      if (true_val && false_val) {
```

Comment:
I'm confused by this logic. We're checking if both values were already defined in the parent scope and if they both are then if both types are equal to each other we allow them unify to union ? If both types are equal to each other, shouldn't they already unify ? 

I'm not sure how whether or not they were defined in the parent scope should matter. Plus, `true_val` should always equal `false_val` since both save_true and save_false have the same parent scopes.

Reply:
Good catch. This was me misunderstanding the API, but I have it fixed now. What I really wanted was to check the type in `environment_stack->type_table_` if it exists. It could be the case that a variable is annotated as being a Union but is refined in the if-else branches, e.g. 

```python
def fn(b: bool):
    y: Union[int, str, None] = None
    if (b):
        # `y` should be a `str` in this branch without needing an
        # `isinstance` check
        y = "foo"
    else:
       # same here
        y = 1
    # But outside of the if-else branche, whatever the value 
    # is (whether `"foo"` or `1`) should go back to being a Union
    # and would need an `isinstance` for type refinement
```

Reply:
Thanks, I was using the wrong function in our API. It should be fixed now, but please take a look at it again just to make sure.